### PR TITLE
Fix M1 Docker Emulation

### LINKED ISSUE #936: Minimize production Docker image
### Is your feature request related to a specific problem?

* There are packages e.g. vim included in the current fidesctl image that are not strictly necessary for the application to function. Aside from image bloat, this is a security risk. An attacker can't exploit what isn't included (attack surface reduction). 

### Describe the solution you'd like

* Identify the essential packages that need to be included in a production image.
* Modify the build process to include strictly essential packages & files in the production image.

### Describe alternatives you've considered, if any

n/a

### Additional context

* This issue also applies to fidesops, but I haven't created a separate issue for it.
* Dev images can be bloated. Indeed, they should be optimized for build speed, not size.
* Multi-stages builds might help, but are of limited use for Python application images.
* Moving from Debian to Alpine as a base image is probably a dead end. Build durations would [mushroom](https://pythonspeed.com/articles/alpine-docker-python/) (alienating devs) and musl libc is [problematic](https://news.ycombinator.com/item?id=10782897) (no wheels).
* [Some basic tips](https://pythonspeed.com/docker/#best-practices-for-production)

### LINKED ISSUE #1051: Remove mssql dependency from default docker build to improve performance
### Is your feature request related to a specific problem?
The fides Docker build includes installation of Microsoft SQL Server dependencies, such as `mssql-tools`. These dependencies increase build time significantly, even though the mssql features are [optional "dangerous" extras](https://github.com/ethyca/fides/blob/6e7c80b565388a535b6eea8adf8b916d41202c85/setup.py#L33-L34). 
[Fidesops](https://github.com/ethyca/fidesops/blob/main/Dockerfile#L37) has already introduced an env var to avoid these dependencies.

These dependencies also force us to use [AMD64 base docker image](https://github.com/ethyca/fides/blob/0f9c9062/Dockerfile#L1), because the packages are published for ARM64:
`FROM --platform=linux/amd64 python:3.9-slim-bullseye as base`

This introduces a dramatic performance degradation on Apple M1 chips, which are common for developer machines. Docker Desktop warns about this problem:

> ⚠️ AMD64
> Image may have poor performance, or fail, if run via emulation.

### Describe the solution you'd like

The main docker build stages should all avoid mssql and the platform restriction: base, builder, dev, and prod.

To support Microsoft SQL users, there are a few options:
1. Documentation on the MSSQL extras and the dependencies required.
    - Perhaps a sample Dockerfile for using fides a base and layers in the platform + dependencies
2. A build stage that includes the platform specifier and the dependency installation
    - Could this image be published with a dedicated tag? 

### Describe alternatives you've considered, if any
- Have the "with mssql" be the default and document how to build "without mssql"
    - This reduces how much the image will change in the next release
    - But mssql usage is not the 90% use case
- Throw M1 chips into the ocean
    - The macbook "professional" doesn't support two+ external displays. This machine is a $1,500 prank.
    - Unfortunately I already keep all my used car batteries in the ocean.

### PR DESCRIPTION
**Pull request by ThomasLaPiana**
Closes #1051 
Closes #936 

### Code Changes

* [x] refactor build stages to minimize image
* [x] add conditional platform checks so that the build works automagically for ARM
* [x] add a `dangerous-requirements.txt` file for `pyodbc`
* [x] add a `--no-cache` posarg for the Nox build commands (`nox -s "build(dev)" -- nocache`)
* [x] let the user force building one or the other platform images
* [x] add a try/except to the importation of mssql-specific modules

### Steps to Confirm

* [ ] someone with an M1 mac confirm that the docker setup works with the new platform trick
* [ ] someone with an M1 mac confirm that build/run speeds are improved
* [ ] docker platform emulation can be forced using posargs for the `build` command (described below)
* [ ] verify that tests can be collected for ctl and ops with the ARM platform (but the mssql tests will fail)

### Pre-Merge Checklist

* [ ] All CI Pipelines Succeeded
* Documentation Updated:
  - [ ] documentation complete, or draft/outline provided (tag docs-team to complete/review on this branch)
  - [ ] documentation issue created (tag docs-team to complete issue separately)
* [x] Issue Requirements are Met
* [ ] Relevant Follow-Up Issues Created
* [ ] Update `CHANGELOG.md`

### Description Of Changes

This PR refactors the Dockerfile to accomplish two goals:
1. Improve the developer experience for anyone on an M1 mac by avoiding the use of platform emulation in the dev environment
2. Minimizing the production image where possible by reducing dependencies and better leveraging efficient build stages

Running certain tests via nox will now fail unless emulation is forced via the build command. This new platform selection has 2 options:
1. Do nothing, and `nox -s build` will automatically choose the right platform for your machine
2. Run the build command with a posarg: `nox -s build -- [amd64 | arm64]`

As an M1 user, forcing the platform with `nox -s build -- amd64` will allow MSSQL tests to pass, but significantly hinder performance. Likewise, for testing, non-M1 users can use `nox -s build -- arm64` to force ARM emulation.

### CODE REVIEW

**Comment by ThomasLaPiana** [comment]
I don't really have a way to test this, but I'm thinking we might also need a platform check here for the mssql portions

**Comment by ThomasLaPiana** [comment]
note here from slack that on M1 machines they come up as `arm64`, vs. `amd64` for x86 architecture

**Comment by ThomasLaPiana** [comment]
figured out that I could test by forcing emulation for arm, now trying to figure out how to skip installing pyodbc

**Comment by ThomasLaPiana** [comment]
all currently failing tests are expected/known occasionally failing

**Comment by sanders41** [comment]
First off in testing this is SOOOO much better 🚀 

Previous build times: ~ 20-30 minutes
Build time now with no cache: ~ 2 minutes

Testing in the container:
Previously ~ 3 minutes for test collection
Test collection now ~ 20 seconds

In testing I'm seeing a lot of test failures that don't seem related to this change, and I don't know why they fail now but didn't before. One thing I have seen is in the unit test nox command it has `-m "not integration and not integration_external and not integration_saas"` and I am having tests with `@pytest.mark.external` fail. So our mark should at least be `-m "not integration and not integration_external and not integration_saas and not external"`. The current mark is what it has always been so I don't know why it worked previously and not here.

Would you rather work on fixing these issues in this PR, or create a separate issue to fix this and any new ones like it that come up?

File: docs/fides/docs/development/overview.md [hunk 1]
```diff
@@ -18,41 +18,37 @@ Once that's complete, there are a few different ways to spin up the project and
...
-    1. Install the `Remote-Containers` extension for VS Code.
-    1. Open the command window (the F1 key will open it, by default) and select the `Remote-Containers: Open Folder in Container...` command.
-    1. Click "Open", without having selected any specific folder.
-    1. The containers will now spin up and VS Code will be running inside of the containers. The bottom left of the IDE will now say `Dev Container: Fides`.
-    1. When you open a new VS Code shell, it will be inside of the `fides` container, and you'll have access to all of the `fides` commands as well as any Python commands like `pytest`, `black`, `mypy`, etc.
-1. If you're using an editor besides VS Code, then the next best way to work on Fides is by utilizing the `Noxfile` commands:
-    1. Make sure that you have `docker`, `docker-compose` and `nox` (`pipx install nox`) installed.
-    1. Once you have everything set up, run `nox -s cli` to spin up a shell within the `fides` container.
+1. The next best way to work on Fides is by utilizing the `Noxfile` commands:
```
Comment (sanders41) [inline docs/fides/docs/development/overview.md:21]:
```suggestion
1. The recommended way to work on Fides is by utilizing the `Noxfile` commands:
```

**Review by sanders41 — CHANGES_REQUESTED** [review]
I am getting some errors on the ops integration tests because the mssql driver isn't installed.

At startup of the testings I get

```console
Traceback (most recent call last):
  File "/fides/tests/ops/integration_tests/setup_scripts/mssql_setup.py", line 39, in <module>
    mssql_setup()
  File "/fides/tests/ops/integration_tests/setup_scripts/mssql_setup.py", line 16, in mssql_setup
    engine = sqlalchemy.create_engine(MASTER_MSSQL_URL)
  File "<string>", line 2, in create_engine
  File "/root/.local/lib/python3.10/site-packages/sqlalchemy/util/deprecations.py", line 309, in warned
    return fn(*args, **kwargs)
  File "/root/.local/lib/python3.10/site-packages/sqlalchemy/engine/create.py", line 560, in create_engine
    dbapi = dialect_cls.dbapi(**dbapi_args)
  File "/root/.local/lib/python3.10/site-packages/sqlalchemy/connectors/pyodbc.py", line 43, in dbapi
    return __import__("pyodbc")
ModuleNotFoundError: No module named 'pyodbc'
```

For this one what I am thinking we could to is change [this line](https://github.com/ethyca/fides/blob/763d605b577cb6b813d258e9f3550c04af67ae99/tests/ops/integration_tests/setup_scripts/mssql_setup.py#L39) to

```py
    try:
        mssql_setup()
    except ModuleNotFoundError:
        print("MS SQL drivers not installed, skipping setup")
```

I also get errors on the integration tests that interact with MS SQL. 

```console
FAILED tests/ops/integration_tests/test_connection_configuration_integration.py::TestMicrosoftSQLServerConnection::test_mssql_db_connection_connect_with_components
FAILED tests/ops/integration_tests/test_connection_configuration_integration.py::TestMicrosoftSQLServerConnection::test_mssql_db_connection_connect_with_url
FAILED tests/ops/integration_tests/test_connection_configuration_integration.py::TestMicrosoftSQLServerConnection::test_connection_configuration_test
FAILED tests/ops/integration_tests/test_connection_configuration_integration.py::TestMicrosoftSQLServerConnection::test_mssql_db_connector - ModuleNotF...
ERROR tests/ops/integration_tests/test_integration_mssql_example.py::test_mssql_example_data - ModuleNotFoundError: No module named 'pyodbc'
ERROR tests/ops/integration_tests/test_sql_task.py::test_mssql_access_request_task - ModuleNotFoundError: No module named 'pyodbc'
ERROR tests/ops/service/privacy_request/request_runner_service_test.py::test_create_and_process_access_request_mssql - ModuleNotFoundError: No module n...
ERROR tests/ops/service/privacy_request/request_runner_service_test.py::test_create_and_process_erasure_request_specific_category_mssql - ModuleNotFoun...
```

Maybe for these we could mark them as `dangerous-integration` rather than `integration` or something like that?

**Comment by ThomasLaPiana** [comment]
Good finds, I agree we'll need some conditional import checking and that looks like the spot to do it!

As for the tests, I'd be ok to add/update that mark, or we can skip the `integration_mssql` ones specifically for now

**Comment by sanders41** [comment]
> As for the tests, I'd be ok to add/update that mark, or we can skip the integration_mssql ones specifically for now

Skipping mssql works also, they are the only ones causing issues.

**Comment by ThomasLaPiana** [comment]
@sanders41 so the solution I decided on here was to add an option to force which platform to build for, instead of splitting out the tests specifically. This allows users to force the platform and run everything as it would in CI.

For local development, I felt that people would be running tests in a more targeted fashion from within their shell, and that the MSSQL tests are such a small fraction of the overall test coverage that this wouldn't be a major hindrance.

This is compounded by the fact that `run_infrastructure` is the one performing all of the docker-level setup (instead of pytest itself, which I think should be the one responsible), so filtering MSSQL out of the rest of the pytest marks and setting it up separately is in and of itself large enough to be its own ticket.

**Comment by ThomasLaPiana** [comment]
i confirmed that all tests can be collected without issue for both `ctl` and `ops` using the emulated dockerfile

**Comment by ThomasLaPiana** [comment]
errors here seem to be unrelated to code changes (external mailchimp test failing)

**Review by sanders41 — APPROVED** [review]
I agree on all points. I would love to get the run infrastructure into pytest.

It also looks to me like these test failures aren't related to this PR so approving.

**Comment by ThomasLaPiana** [comment]
appreciate the review @sanders41 I know this was a heavy one!

**Pull request merged**